presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/SimplifyExpressions.java

```diff
@@ -212,11 +213,29 @@ public Expression rewriteExpression(Expression node, NodeContext context, Expres
         @Override
         public Expression rewriteLogicalBinaryExpression(LogicalBinaryExpression node, NodeContext context, ExpressionTreeRewriter<NodeContext> treeRewriter)
         {
-            List<Expression> predicates = extractPredicates(node.getType(), node).stream()
-                    .map(expression -> treeRewriter.rewrite(expression, NodeContext.NOT_ROOT_NODE))
-                    .collect(toImmutableList());
+            Expression expression = combinePredicates(
+                    node.getType(),
+                    extractPredicates(node.getType(), node).stream()
+                            .map(subExpression -> treeRewriter.rewrite(subExpression, NodeContext.NOT_ROOT_NODE))
+                            .collect(toImmutableList()));
+
+            if (!(expression instanceof LogicalBinaryExpression)) {
+                return expression;
+            }
+
+            Expression simplified = extractCommonPredicates((LogicalBinaryExpression) expression);
+
+            // Prefer AND LogicalBinaryExpression at the root if possible
+            if (context.isRootNode() && simplified instanceof LogicalBinaryExpression && ((LogicalBinaryExpression) simplified).getType() == OR) {
+                return distributeIfPossible((LogicalBinaryExpression) simplified);
```

Comment:
Yes, it seems to be handled correctly since first `X` and `Y` are `OR-ed` and then `AND-ed` as only cross product.